Source/WebCore/Modules/webcodecs/WebCodecsAudioDataAlgorithms.cpp

```diff
@@ -0,0 +1,132 @@
+        return false;
+
+    if (!init.numberOfChannels)
+        return false;
+
+    auto totalSamples = init.numberOfFrames * init.numberOfChannels;
+    auto bytesPerSample = computeBytesPerSample(init.format);
+    auto totalSize = totalSamples * bytesPerSample;
+    auto dataSize = init.data.span().size();
+    return dataSize >= totalSize;
+}
+
+bool isAudioSampleFormatInterleaved(const AudioSampleFormat& format)
+{
+    switch (format) {
+    case AudioSampleFormat::U8:
+    case AudioSampleFormat::S16:
+    case AudioSampleFormat::S32:
+    case AudioSampleFormat::F32:
+        return true;
+    case AudioSampleFormat::U8Planar:
+    case AudioSampleFormat::S16Planar:
+    case AudioSampleFormat::S32Planar:
+    case AudioSampleFormat::F32Planar:
+        return false;
+    }
+    RELEASE_ASSERT_NOT_REACHED();
+    return false;
+}
+
+size_t computeBytesPerSample(const AudioSampleFormat& format)
+{
+    switch (format) {
+    case AudioSampleFormat::U8:
+    case AudioSampleFormat::U8Planar:
+        return 1;
+    case AudioSampleFormat::S16:
+    case AudioSampleFormat::S16Planar:
+        return 2;
+    case AudioSampleFormat::S32:
+    case AudioSampleFormat::F32:
+    case AudioSampleFormat::S32Planar:
+    case AudioSampleFormat::F32Planar:
+        return 4;
+    }
+    RELEASE_ASSERT_NOT_REACHED();
+    return 0;
+}
+
+// https://www.w3.org/TR/webcodecs/#compute-copy-element-count
+ExceptionOr<size_t> computeCopyElementCount(const WebCodecsAudioData& data, const WebCodecsAudioData::CopyToOptions& options)
+{
+    auto destFormat = options.format.value_or(*data.format());
+    auto isInterleaved = isAudioSampleFormatInterleaved(destFormat);
+    if (isInterleaved && options.planeIndex > 0)
+        return Exception { RangeError, "Invalid planeIndex for interleaved format"_s };
+    if (options.planeIndex >= data.numberOfChannels())
+        return Exception { RangeError, "Invalid planeIndex for planar format"_s };
+
+    if (options.format && *options.format != destFormat && destFormat != AudioSampleFormat::F32Planar)
+        return Exception { NotSupportedError, "AudioData currently only supports copy conversion to f32-planar"_s };
+
+    auto platformData = data.data().audioData;
+    if (!platformData)
+        return Exception { InvalidAccessError, "Internal AudioData storage is null"_s };
+
+    auto frameCount = data.numberOfFrames() / data.numberOfChannels();
+    if (options.frameOffset && *options.frameOffset > frameCount)
+        return Exception { RangeError, "frameOffset is too large"_s };
+
+    auto copyFrameCount = frameCount;
+    if (options.frameOffset)
+        copyFrameCount -= *options.frameOffset;
+    if (options.frameCount) {
+        if (*options.frameCount > copyFrameCount)
+            return Exception { RangeError, "frameCount is too large"_s };
+
+        copyFrameCount = *options.frameCount;
+    }
+
+    if (isInterleaved)
+        copyFrameCount *= data.numberOfChannels();
```

Comment:
Ditto for these multiplications, maybe we should use safe operations and throw if it overflows.